### ISSUE TITLE
Testsuite fails when rebuilding using cache

### ISSUE THREAD

**Issue by snowleopard** [ocaml/dune#3376]
While working on #3370, I noticed that Dune cache doesn't interact well with tests. Below is a way to reproduce. Don't forget to enable Dune cache by adding `(cache enabled)` to Dune config.

```bash
$ rm -rf /home/amokhov/.cache/dune/db
$ rm -rf _build
$ make dev
$ _build/default/bin/main.exe runtest
$ cp _build/default/bin/main.exe the_dune.exe
$ rm -rf _build
$ ./the_dune.exe runtest
```

This results in multiple failures, all due to file permissions. For example:

```
$ echo "(lang dune 2.0)" > dune-project
+  /tmp/dune-test6f4569.sh: line 1: dune-project: Permission denied
+  [1]
```

@mefyl Is this a known issue? Any ideas on what could be happening here?

I found another issue about the interaction of the testsuite and Dune cache: #2802. It appears to be unrelated though.

**Comment by snowleopard** [comment]
When running `_build/default/bin/main.exe build @test/blackbox-tests/duplicate-c-cxx-obj`, we end up promoting a bunch of targets to the cache, including `dune-project` that appears in the above error message, which causes the files to be set read-only.

It is unclear, however, why these files appear in the list of targets. The corresponding alias doesn't seem to declare any targets:

```
(rule
 (alias duplicate-c-cxx-obj)
 (deps (package dune) (source_tree test-cases/duplicate-c-cxx-obj))
 (action
  (chdir
   test-cases/duplicate-c-cxx-obj
   (progn (run dune-cram run run.t) (diff? run.t run.t.corrected)))))
```

**Comment by snowleopard** [comment]
Together with @aalekseyev we figured out the problem. In the test cases that fail, we have a few source files like `dune` and `dune-project`. Dune first copies them to the build directory, and this step happens to be cached. But then, when running tests we want to modify these files, and fail over because the files have been cached and declared read-only.

To fix things, I will modify the tests, removing source files that are modified during test, and creating them in `run.t` instead.

**Comment by snowleopard** [comment]
Maybe Dune should set source files as read-only when copying them to the build directory, regardless of whether Dune cache is enabled or not. This would allow us to catch such issues earlier.

**Comment by snowleopard** [comment]
Looks like marking targets read-only has already been implemented in #3092, and all we need to do is to bump the version in Dune's `dune-project` file.
